theano/compile/function_module.py

```diff
@@ -483,6 +488,66 @@ def __copy__(self):
                 there.data = here.data
         return cpy
 
+    def copy(self, share_memory = False):
+        """
+        Copy this function. Copied function will have separated maker and fgraph
+        with original function. User can choose whether to separate storage by 
+        changing the share_memory arguments
+        ---------------------
+        Params:
+            share_memory -- { boolean } Default is False. When True, two function
+            share intermediate storages(storages except input and output storages)
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+
+        if not share_memory:
+            return self.__copy__()
+        else:
+            # copy SymbolocKits
+            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
```

Comment:
In() containt reference to the variable used before creating the function
graph. When we create the fgraph, we clone the user graph as we do not want
to modify the user graph itself. But we never up to now made the In() point
to the cloned var. So you do not need to do this.

To know the mapping between maker.inputs and fgraph.inputs, the order is
the same. Both are lists.

There is the notion of extended inputs and extended outputs. They come by
appending the shared var used to the inputs and the update to the outputs.

The function graph only see the extended inputs and outputs. FunctionMaker
also only see the extended inputs and outputs.

When you call theano.function() if you use shared variable, this will call
pfunc(). It is him that will create the extended inputs list and call
orig_function() with it. orig_function, call std_fgraph() and it is him
that create the extended outputs that are passed to the function graph.

So f.maker.inputs contain the extended inputs, but not the extended outputs.

f.maker.fgraph.inputs/outputs contain a cloned copy of the extended inputs
and outputs.

Yes this is ugly. I needed to dig in the code to tell you this. A clean up
of that would be good, but not for now. Can you take notes of what would
need clean up and what could be done?

On Thu, May 28, 2015 at 9:07 AM, Chienli Ma(马千里) <notifications@github.com>
wrote:

> In theano/compile/function_module.py
> <https://github.com/Theano/Theano/pull/2969#discussion_r31230495>:
>
> > +        with original function. User can choose whether to separate storage by
> > +        changing the share_memory arguments
> > +        ---------------------
> > +        Params:
> > +            share_memory -- { boolean } Default is False. When True, two function
> > +            share intermediate storages(storages except input and output storages)
> > +        ---------------------
> > +        Returns:
> > +            func -- Copied theano.Function
> > +        """
> > +
> > +        if not share_memory:
> > +            return self.__copy__()
> > +        else:
> > +            # copy SymbolocKits
> > +            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
>
> Well, I come out with a new approach and I am working on it. But I still
> don't understand why variables in In() don't refer to same object to
> those in fgraph. Does it matters?. Also I don't know how the update work --
> I could not find the codes. I would like to know Any hints?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/2969/files#r31230495>.
>